### PR TITLE
Move email sending logic out of transaction

### PR DESCRIPTION
**Pull request by allenwq**
Follow #1028, there are still some failure specs.

### CODE REVIEW

File: app/services/course/user_registration_service.rb [hunk 1]
```diff
@@ -5,14 +5,16 @@ class Course::UserRegistrationService
   # @param [Course::Registration] registration The registration object to be processed.
   # @return [Boolean] True if the registration succeeded. False if the registration failed.
   def register(registration)
+    course_user = nil
```
Comment (lowjoel) [inline app/services/course/user_registration_service.rb]:
```ruby
course_user = CourseUser.trasnaction { create_or_update_registration(registration) }
```

File: app/services/course/user_registration_service.rb [hunk 2]
```diff
@@ -5,14 +5,16 @@ class Course::UserRegistrationService
...
-      succeeded = course_user && !course_user.changed?
-      if succeeded && course_user.requested?
-        notify_course_staff(registration.course, course_user)
-      else
-        succeeded
-      end
+    end
+
+    succeeded = course_user && !course_user.changed?
+    if succeeded && course_user.requested?
```
Comment (lowjoel) [inline app/services/course/user_registration_service.rb]: maybe

```ruby
return succeeded unless course_user.requested?
notify_course_staff(registration.course, course_user)
```

File: app/services/course/user_registration_service.rb (hunk 1 above)
Comment (allenwq) [inline app/services/course/user_registration_service.rb]: 👍

File: app/services/course/user_registration_service.rb [hunk 3]
```diff
@@ -5,15 +5,12 @@ class Course::UserRegistrationService
...
-      if succeeded && course_user.requested?
-        notify_course_staff(registration.course, course_user)
-      else
-        succeeded
-      end
-    end
+    course_user = CourseUser.transaction { create_or_update_registration(registration) }
+
+    succeeded = course_user && !course_user.changed?
+    return succeeded unless course_user.requested?
```
Comment (lowjoel) [inline app/services/course/user_registration_service.rb]: urgh, crap. this changed the meaning, sorry.

Comment (allenwq) [inline app/services/course/user_registration_service.rb]: I don't know course_user can be nil either > <  I guess that's why it was implemented like that before.

**Pull request merged**